Change minlex to use the linear time least_rotation.

MarkCBell opened the pull request:
Currently the `minlex` utility function computes the earliest rotation of an iterable by producing all rotations of the iterable and looking for the smallest one. This is a quadratic time operation since there are linearly many rotations and it takes linear time to determine whether a rotation is smaller than the best rotation found so far. The utility function `least_rotation` uses Booth's algorithm to compute the index of the smallest rotation in linear time hence `minlex` can be fundamentally done using:
``` return left_rotate(seq, least_rotation(seq))```
however a few extra cases are needed since `minlex` can also check for the smallest rotation of the iterable and its reverse.

Since this is now a linear time algorithm there is no benefit in passing in some of the hints that speed up the old function such as `is_set` and `small`.

<!-- BEGIN RELEASE NOTES -->
* utilities
    * minlex no longer accepts is_set or small arguments
    * minlex and least_rotation now accept `key=` arguments similar to `sorted`.
<!-- END RELEASE NOTES -->

MarkCBell committed:
Changed minlex to use the linear time least_rotation.

smichr commented:
This appears to no longer retain the string/tuple invariant for output.

MarkCBell committed:
Minlex returns tuple or string, docstring updated too.

MarkCBell commented:
> This appears to no longer retain the string/tuple invariant for output.

Thank you for letting me know about this. I have added a commit to ensure that the returned type is a tuple, unless the input is a string and have updated the docstring of the function to note that this is the desired behaviour.

smichr commented:
It would be good to add a test for a string, too.

smichr commented on sympy/utilities/iterables.py:
```diff
@@ -2430,50 +2426,15 @@ def minlex(seq, directed=True, is_set=False, small=None):
+    if not directed:
+        rseq = seq[::-1]
+        rbest = rotate_left(rseq, least_rotation(rseq))
+        best = min(best, rbest, key=default_sort_key)
```
This might be more restrictive. Things in iterables should mostly be sympy-agnostic. So previously the items were simply compared: `if seq < best`

smichr commented:
Using "default_sort_key" might be more restrictive than this was before. Things in iterables should mostly be sympy-agnostic. So previously the items were simply compared: `if seq < best`

oscarbenjamin commented:
What is the reason for using `default_sort_key`?

MarkCBell commented:
So without the default sort key the Polyhedron test (`py.test -k test_args`) fails since it tries to sort a bunch of Relationals which are not naturally orderable. The old version of this function got around this by using the `default_sort_key` method.

```
________________ test_sympy__combinatorics__polyhedron__Polyhedron _________________

    def test_sympy__combinatorics__polyhedron__Polyhedron():
        from sympy.combinatorics.permutations import Permutation
        from sympy.combinatorics.polyhedron import Polyhedron
        from sympy.abc import w, x, y, z
        pgroup = [Permutation([[0, 1, 2], [3]]),
                  Permutation([[0, 1, 3], [2]]),
                  Permutation([[0, 2, 3], [1]]),
                  Permutation([[1, 2, 3], [0]]),
                  Permutation([[0, 1], [2, 3]]),
                  Permutation([[0, 2], [1, 3]]),
                  Permutation([[0, 3], [1, 2]]),
                  Permutation([[0, 1, 2, 3]])]
        corners = [w, x, y, z]
        faces = [(w, x, y), (w, y, z), (w, z, x), (x, y, z)]
>       assert _test_args(Polyhedron(corners, faces, pgroup))

sympy/core/tests/test_args.py:514: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/combinatorics/polyhedron.py:389: in __new__
    faces = [minlex(f, directed=False) for f in faces]
sympy/combinatorics/polyhedron.py:389: in <listcomp>
    faces = [minlex(f, directed=False) for f in faces]
sympy/utilities/iterables.py:2430: in minlex
    best = rotate_left(seq, least_rotation(seq))
sympy/utilities/iterables.py:1309: in least_rotation
    if (sj) < (S[k]):

self = x < w

    def __bool__(self):
>       raise TypeError("cannot determine truth value of Relational")
E       TypeError: cannot determine truth value of Relational

sympy/core/relational.py:395: TypeError
```

oscarbenjamin commented:
Sorry, I didn't see that `default_sort_key` was also being used in the old implementation.

@smichr does this look good to you?

(I don't know this particular routine so well but the change seems reasonable to me from a high level)

MarkCBell commented:
So having given this some thought, I think one refinement that would be reasonable is to adjust both `minlex` and ` least_rotation` have a `key=lambda x: x` argument which is then used on the comparisons. It would then be up to the `Polyhedron` constructor to use `minlex` with `key=default_sort_key`, that is `minlex(f, directed=False, key=default_sort_key)` [here](https://github.com/sympy/sympy/pull/20433/files#diff-627798760f667621e0acdede3c95c236fd83d99f71e87ba205791afffd5851e1L389 )

MarkCBell commented:
> It would be good to add a test for a string, too.

@smichr There is already a doctest for this. Are you suggesting a separate unittest too?

oscarbenjamin commented:
Doctests are just for testing the documentation. We should have unit tests to test the correctness of the code.

MarkCBell commented:
> Doctests are just for testing the documentation. We should have unit tests to test the correctness of the code.
Ok, I've just checked and [there is also a unittest already](https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_iterables.py#L647).

MarkCBell committed:
Added key= argument to minlex and least_rotation.

MarkCBell committed:
Ignoring flake8 unused import.

A future pull request could replace all uses of:
    from sympy.utilities.iterables import default_sort_key
and:
    from sympy.utilities import default_sort_key
with:
    from sympy.core.compatibility import default_sort_key

oscarbenjamin commented:
I thought it was ready to merge. What were the last two commits for?

MarkCBell commented:
>I thought it was ready to merge. What were the last two commits for?

So following the discussion around whether `least_rotation` should use `default_sort_key` I have replaced this with a generic `key` function (which is `sympy.Id` by default). Users can then pass in `key=default_sort_key` if required, and this is now what is done within the Polyhedron method which (unusually) requires this.

I've added a unittest  & doctest to demonstrate the use of this.

oscarbenjamin commented:
Thanks @MarkCBell and @smichr! This looks good
